File: src/lib/generalCategories.ts
export type GeneralCategoryDirection = 'income' | 'expense';

export type GeneralCategory = {
  name: string;
  icon: string;
  direction: GeneralCategoryDirection;
};

// —— 预设类别 ——
export const GENERAL_EXPENSE_CATEGORIES: GeneralCategory[] = [
  { name: '餐饮', icon: '🍜', direction: 'expense' },
  { name: '交通', icon: '🚌', direction: 'expense' },
  { name: '购物', icon: '🛍️', direction: 'expense' },
  { name: '居住', icon: '🏠', direction: 'expense' },
  { name: '娱乐', icon: '🎬', direction: 'expense' },
  { name: '医疗', icon: '💊', direction: 'expense' },
  { name: '人情', icon: '🎁', direction: 'expense' },
  { name: '学习', icon: '📚', direction: 'expense' },
  { name: '其它支出', icon: '💸', direction: 'expense' },
];

export const GENERAL_INCOME_CATEGORIES: GeneralCategory[] = [
  { name: '工资', icon: '💰', direction: 'income' },
  { name: '奖金', icon: '🎊', direction: 'income' },
  { name: '副业', icon: '💡', direction: 'income' },
  { name: '理财', icon: '📈', direction: 'income' },
  { name: '红包', icon: '🧧', direction: 'income' },
  { name: '其它收入', icon: '💵', direction: 'income' },
];

export const ALL_GENERAL_CATEGORIES: GeneralCategory[] = [
  ...GENERAL_EXPENSE_CATEGORIES,
  ...GENERAL_INCOME_CATEGORIES,
];

// —— 图标库（用于自定义类别选择）——
export const ICON_LIBRARY: { group: string; icons: string[] }[] = [
  {
    group: '餐饮',
    icons: ['🍜', '🍚', '🍱', '🍔', '🍕', '🍣', '🍰', '☕', '🥤', '🍺', '🍷', '🍎', '🍇', '🥗', '🍳', '🥟'],
  },
  {
    group: '交通',
    icons: ['🚌', '🚗', '🚕', '🚇', '✈️', '🚄', '🛵', '🚲', '⛽', '🅿️', '🛴', '🚢'],
  },
  {
    group: '购物',
    icons: ['🛍️', '👗', '👟', '💄', '💍', '👜', '📱', '💻', '🎧', '⌚', '🧸', '🎮'],
  },
  {
    group: '居家',
    icons: ['🏠', '🛏️', '🛋️', '🍳', '🧴', '🧻', '💡', '🔧', '🧹', '🪴', '📦'],
  },
  {
    group: '娱乐',
    icons: ['🎬', '🎵', '🎤', '🎨', '🎭', '🎳', '🎲', '🎯', '🎪', '🏖️', '🌊', '⛰️', '🎢', '🎡', '🕹️'],
  },
  {
    group: '健康',
    icons: ['💊', '🏥', '🩺', '🦷', '👓', '🏋️', '🧘', '🚴', '⚽', '🏀', '🎾', '🧴'],
  },
  {
    group: '人情',
    icons: ['🎁', '💐', '🎂', '🎉', '💒', '👶', '🧧', '💌', '🥂'],
  },
  {
    group: '学习',
    icons: ['📚', '✏️', '🎓', '📖', '📝', '🎒', '🖥️', '📊', '🔬', '🎨'],
  },
  {
    group: '宠物',
    icons: ['🐶', '🐱', '🐰', '🐹', '🐦', '🐟', '🦴', '🥩'],
  },
  {
    group: '通讯',
    icons: ['📱', '📞', '📶', '📡', '💬', '📧', '📮'],
  },
  {
    group: '收入',
    icons: ['💰', '💵', '💴', '💶', '💷', '💳', '🏦', '📈', '💡', '🎊', '🧧', '🏆', '💎'],
  },
  {
    group: '其它',
    icons: ['💸', '📝', '🌟', '⭐', '❤️', '🔥', '⚡', '☔', '☀️', '🌙', '🌈', '🎯', '📌', '🔔'],
  },
];

// —— 用户自定义（存于 Ledger.customCategories JSON）——
export type CustomCategoriesJson = {
  added: GeneralCategory[];
  hidden: string[]; // 被隐藏的类别名（可含预设 or 自定义）
  // 分类别月预算：{ [category name]: cents }。老账本没有这个字段，parseCustom 兜底成空。
  // 与 Ledger.budgetCents（账本级总预算）并存 —— 总预算是"这个月最多花多少"，
  // 分类预算是"某类最多花多少"，两者独立展示，不做联动校验（用户可能故意让分类
  // 之和超过总预算，比如给"其它支出"留缓冲）
  budgets?: Record<string, number>;
};

export function parseCustom(json: string | null | undefined): CustomCategoriesJson {
  if (!json) return { added: [], hidden: [], budgets: {} };
  try {
    const p = JSON.parse(json);
    const budgets =
      p.budgets && typeof p.budgets === 'object' && !Array.isArray(p.budgets)
        ? Object.fromEntries(
            Object.entries(p.budgets).filter(
              ([, v]) => typeof v === 'number' && Number.isFinite(v) && v > 0,
            ) as [string, number][],
          )
        : {};
    return {
      added: Array.isArray(p.added) ? p.added : [],
      hidden: Array.isArray(p.hidden) ? p.hidden : [],
      budgets,
    };
  } catch {
    return { added: [], hidden: [], budgets: {} };
  }
}

export function stringifyCustom(c: CustomCategoriesJson): string {
  // 空 budgets 不落库 —— 让 diff 干净，也让"没有分类预算"和"预算全清 0"两种状态一致
  const clean: CustomCategoriesJson = {
    added: c.added,
    hidden: c.hidden,
  };
  if (c.budgets && Object.keys(c.budgets).length > 0) clean.budgets = c.budgets;
  return JSON.stringify(clean);
}

/** 取某类别的月预算（分）；无则 null */
export function categoryBudgetOf(
  customJson: string | null | undefined,
  category: string,
): number | null {
  const budgets = parseCustom(customJson).budgets ?? {};
  const v = budgets[category];
  return typeof v === 'number' && v > 0 ? v : null;
}

// —— 合并后的有效类别（预设 - 隐藏 + 新增）——
export function effectiveCategories(
  customJson: string | null | undefined,
  direction: GeneralCategoryDirection,
): GeneralCategory[] {
  const { added, hidden } = parseCustom(customJson);
  const presets =
    direction === 'expense' ? GENERAL_EXPENSE_CATEGORIES : GENERAL_INCOME_CATEGORIES;
  const hiddenSet = new Set(hidden);
  const kept = presets.filter((c) => !hiddenSet.has(c.name));
  const addedThis = added.filter((c) => c.direction === direction && !hiddenSet.has(c.name));
  // 去重（自定义名如果撞了预设名以自定义为准）
  const nameSet = new Set(kept.map((c) => c.name));
  const uniqAdded = addedThis.filter((c) => !nameSet.has(c.name));
  return [...kept, ...uniqAdded];
}

export function effectiveAll(customJson: string | null | undefined): GeneralCategory[] {
  return [
    ...effectiveCategories(customJson, 'expense'),
    ...effectiveCategories(customJson, 'income'),
  ];
}

// —— 查找 icon / 方向 ——
export function iconOf(name: string, customJson?: string | null): string {
  const all = customJson ? effectiveAll(customJson) : ALL_GENERAL_CATEGORIES;
  return all.find((c) => c.name === name)?.icon ?? '📝';
}

export function directionOfGeneral(
  name: string,
  customJson: string | null | undefined,
  fallback: GeneralCategoryDirection = 'expense',
): GeneralCategoryDirection {
  const all = customJson ? effectiveAll(customJson) : ALL_GENERAL_CATEGORIES;
  const p = all.find((c) => c.name === name);
  return p ? p.direction : fallback;
}

File: mobile/lib/core/general_categories.dart
import 'dart:convert';

/// 自定义类别（对齐网页端 src/lib/generalCategories.ts 的 GeneralCategory）。
///
/// 服务端 `customCategories.added` 要求是 `{name, icon, direction}` 对象数组
/// （见 /api/ledgers/[id] 的 zod schema），故移动端也用对象存储，才能与网页端
/// 双向兼容：网页端建的带 emoji 自定义类别能在端上显示图标，端上建的也能在
/// 网页端正确归入收入/支出分组。[#1]
class GeneralCategory {
  final String name;
  final String icon;
  final String direction; // 'income' | 'expense'

  const GeneralCategory({
    required this.name,
    this.icon = '🌟',
    this.direction = 'expense',
  });

  Map<String, dynamic> toJson() =>
      {'name': name, 'icon': icon, 'direction': direction};

  /// 从任意 JSON 值构造：新格式为对象；兼容老端上的「纯字符串名」格式
  /// （无 icon/direction，兜底 🌟 / expense）。
  static GeneralCategory fromAny(dynamic v) {
    if (v is Map) {
      final name = (v['name'] ?? '').toString().trim();
      final icon = (v['icon'] ?? '').toString().trim();
      final dir = (v['direction'] ?? 'expense').toString();
      return GeneralCategory(
        name: name,
        icon: icon.isEmpty ? '🌟' : icon,
        direction: dir == 'income' ? 'income' : 'expense',
      );
    }
    return GeneralCategory(name: v.toString().trim());
  }
}

/// 普通账本分类预算解析（对齐网页端 src/lib/generalCategories.ts 的 parseCustom）。
///
/// 账本 [Ledger.customCategories] 存一段 JSON：
/// ```json
/// { "added":[{"name":"健身","icon":"🏋️","direction":"expense"}], "hidden":[...],
///   "budgets":{ "餐饮": 50000 }, "budgetsWeekly":{ "餐饮": 12000 } }
/// ```
/// 其中 budgets / budgetsWeekly 是分类别预算（分）。老账本可能没有这两个字段，
/// 兜底成空 map。
class CustomCategories {
  final List<GeneralCategory> added;
  final List<String> hidden;
  final Map<String, int> budgets;
  final Map<String, int> budgetsWeekly;

  const CustomCategories({
    this.added = const [],
    this.hidden = const [],
    this.budgets = const {},
    this.budgetsWeekly = const {},
  });

  /// 解析；异常或空串都兜底成全空（与网页端一致：不崩、不影响展示）。
  static CustomCategories parse(String? json) {
    if (json == null || json.isEmpty) return const CustomCategories();
    try {
      final p = jsonDecode(json);
      if (p is! Map) return const CustomCategories();
      return CustomCategories(
        added: p['added'] is List
            ? (p['added'] as List)
                .map(GeneralCategory.fromAny)
                .where((c) => c.name.isNotEmpty)
                .toList()
            : const [],
        hidden: p['hidden'] is List
            ? (p['hidden'] as List).map((e) => e.toString()).toList()
            : const [],
        budgets: _parseBudgetMap(p['budgets']),
        budgetsWeekly: _parseBudgetMap(p['budgetsWeekly']),
      );
    } catch (_) {
      return const CustomCategories();
    }
  }

  /// 序列化回 JSON（分类管理页保存时用）。added 落成对象数组，与服务端 schema 一致。
  Map<String, dynamic> toJson() => {
        'added': added.map((c) => c.toJson()).toList(),
        'hidden': hidden,
        'budgets': budgets,
        'budgetsWeekly': budgetsWeekly,
      };
}

/// 解析分类别预算 map（值为正数分）；异常或空都兜底成空 map。
Map<String, int> _parseBudgetMap(dynamic v) {
  final map = <String, int>{};
  if (v is Map) {
    for (final e in v.entries) {
      final val = e.value;
      if (val is num && val.isFinite && val > 0) {
        map[e.key.toString()] = val.toInt();
      }
    }
  }
  return map;
}

/// 默认（内置）分类清单（1:1 对齐网页端常用分类）。
const List<String> defaultCategories = <String>[
  '餐饮',
  '交通',
  '购物',
  '娱乐',
  '居家',
  '居住',
  '医疗',
  '教育',
  '通讯',
  '工资',
  '红包',
  '理财',
  '旅行',
  '其他',
];

/// 分类 -> emoji 图标（1:1 对齐网页端分类图标）。
const Map<String, String> _categoryIcons = <String, String>{
  '餐饮': '🍜',
  '早餐': '🥐',
  '午餐': '🍱',
  '晚餐': '🍲',
  '交通': '🚌',
  '打车': '🚕',
  '地铁': '🚇',
  '购物': '🛍️',
  '服饰': '👕',
  '数码': '📱',
  '居家': '🧺',
  '居住': '🏠',
  '房租': '🏠',
  '水电': '💡',
  '娱乐': '🎮',
  '运动': '🏀',
  '旅行': '✈️',
  '医疗': '💊',
  '教育': '📚',
  '书籍': '📖',
  '通讯': '📞',
  '工资': '💰',
  '收入': '💵',
  '奖金': '🎉',
  '红包': '🧧',
  '理财': '📈',
  '投资': '📊',
  '礼金': '🎁',
  '社交': '🍻',
  '宠物': '🐾',
  '美容': '💄',
  '烟酒': '🍺',
  '孩子': '🍼',
  '捐赠': '🤝',
  '其他': '📦',
};

/// 返回某分类的展示图标。
///
/// 优先用账本自定义类别里保存的 emoji（[customJson] 传账本 customCategories 时），
/// 其次内置分类图标表，最后回退 📦。这样端上/网页端新建的带 emoji 自定义类别，
/// 在列表、分类预算、记账条目里都能显示正确图标。[#1]
String iconOf(String category, [String? customJson]) {
  if (customJson != null && customJson.isNotEmpty) {
    for (final c in CustomCategories.parse(customJson).added) {
      if (c.name == category && c.icon.isNotEmpty) return c.icon;
    }
  }
  return _categoryIcons[category] ?? '📦';
}

/// 自定义类别可选 emoji 图标库（1:1 对齐网页端 generalCategories.ts 的 ICON_LIBRARY）。
///
/// 分组供「新增类别」时按类目浏览选择，覆盖餐饮/交通/购物/居家/娱乐/健康/人情/
/// 学习/宠物/通讯/收入/其它等常用图标。
const List<({String group, List<String> icons})> iconLibrary = [
  (
    group: '餐饮',
    icons: ['🍜', '🍚', '🍱', '🍔', '🍕', '🍣', '🍰', '☕', '🥤', '🍺', '🍷', '🍎', '🍇', '🥗', '🍳', '🥟'],
  ),
  (
    group: '交通',
    icons: ['🚌', '🚗', '🚕', '🚇', '✈️', '🚄', '🛵', '🚲', '⛽', '🅿️', '🛴', '🚢'],
  ),
  (
    group: '购物',
    icons: ['🛍️', '👗', '👟', '💄', '💍', '👜', '📱', '💻', '🎧', '⌚', '🧸', '🎮'],
  ),
  (
    group: '居家',
    icons: ['🏠', '🛏️', '🛋️', '🧴', '🧻', '💡', '🔧', '🧹', '🪴', '📦'],
  ),
  (
    group: '娱乐',
    icons: ['🎬', '🎵', '🎤', '🎨', '🎭', '🎳', '🎲', '🎯', '🎪', '🏖️', '🌊', '⛰️', '🎢', '🎡', '🕹️'],
  ),
  (
    group: '健康',
    icons: ['💊', '🏥', '🩺', '🦷', '👓', '🏋️', '🧘', '🚴', '⚽', '🏀', '🎾'],
  ),
  (
    group: '人情',
    icons: ['🎁', '💐', '🎂', '🎉', '💒', '👶', '🧧', '💌', '🥂'],
  ),
  (
    group: '学习',
    icons: ['📚', '✏️', '🎓', '📖', '📝', '🎒', '🖥️', '📊', '🔬'],
  ),
  (
    group: '宠物',
    icons: ['🐶', '🐱', '🐰', '🐹', '🐦', '🐟', '🦴', '🥩'],
  ),
  (
    group: '通讯',
    icons: ['📞', '📶', '📡', '💬', '📧', '📮'],
  ),
  (
    group: '收入',
    icons: ['💰', '💵', '💴', '💶', '💷', '💳', '🏦', '📈', '🎊', '🏆', '💎'],
  ),
  (
    group: '其它',
    icons: ['💸', '🌟', '⭐', '❤️', '🔥', '⚡', '☔', '☀️', '🌙', '🌈', '📌', '🔔'],
  ),
];
